include/cuco/detail/static_map.inl

```diff
@@ -422,6 +422,83 @@ __device__ bool static_map<Key, Value, Scope, Allocator>::device_mutable_view::i
   }
 }
 
+template <typename Key, typename Value, cuda::thread_scope Scope, typename Allocator>
+template <typename Hash, typename KeyEqual>
+__device__
+  thrust::pair<typename static_map<Key, Value, Scope, Allocator>::device_mutable_view::iterator,
+               bool>
+  static_map<Key, Value, Scope, Allocator>::device_mutable_view::fetch_insert(
+    value_type const& insert_pair, Hash hash, KeyEqual key_equal) noexcept
+{
+  auto current_slot{initial_slot(insert_pair.first, hash)};
+
+  while (true) {
+    key_type const existing_key = current_slot->first.load(cuda::std::memory_order_relaxed);
+    // The user provide `key_equal` can never be used to compare against `empty_key_sentinel` as the
+    // sentinel is not a valid key value. Therefore, first check for the sentinel
+    auto const slot_is_available =
+      detail::bitwise_compare(existing_key, this->get_empty_key_sentinel()) or
+      detail::bitwise_compare(existing_key, this->get_erased_key_sentinel());
+
+    // the key we are trying to insert is already in the map, so we return with failure to insert
+    if (not slot_is_available and key_equal(existing_key, insert_pair.first)) {
+      // If we cannot use a single CAS operation, ensure that the write to
+      // the value part also took place.
+      if constexpr (not cuco::detail::is_packable<value_type>()) {
+        auto& slot_value       = current_slot->second;
+        auto const empty_value = this->get_empty_value_sentinel();
+        while (
+          detail::bitwise_compare(slot_value.load(cuda::std::memory_order_relaxed), empty_value)) {
+          // spin
+        }
+      }
+
+      return thrust::make_pair(current_slot, false);
+    }
+
+    if (slot_is_available) {
+      auto const status = [&]() {
+        // One single CAS operation if `value_type` is packable
+        if constexpr (cuco::detail::is_packable<value_type>()) {
+          return packed_cas(current_slot, insert_pair, key_equal, existing_key);
+        }
+
+        if constexpr (not cuco::detail::is_packable<value_type>()) {
+          // Only use cas_dependent_write; for back_to_back_cas we cannot
+          // guarantee that we get a valid iterator.
+          return cas_dependent_write(current_slot, insert_pair, key_equal, existing_key);
```

Comment:
This always selects the `cas_dependent_write` strategy.

Maybe I missed the relevant section in our discussion but why do we eliminate the `back_to_back_cas` entirely for this function?

Reply:
It's alluded to by the commit right above: We cannot guarantee a valid iterator with the `back_to_back_cas`. It's also been discussed here: https://github.com/NVIDIA/cuCollections/issues/200#issuecomment-1205392853

In short, the problematic case is if two threads want to insert the same key and one gets the key, the other gets the value. Now if a third thread comes along, it will see the key it wants and a non-empty value. However, the second thread that got the value, will first reset it to the empty sentinel first to signal that the first thread can write its value.

Reply:
I've expanded the comment.